actpool/actpool.go

```diff
@@ -473,73 +436,62 @@ func (ap *actPool) removeInvalidActs(acts []action.SealedEnvelope) {
 			continue
 		}
 		log.L().Debug("Removed invalidated action.", log.Hex("hash", hash[:]))
-		delete(ap.allActions, hash)
+		ap.allActions.Delete(hash)
 		intrinsicGas, _ := act.IntrinsicGas()
-		ap.subGasFromPool(intrinsicGas)
+		atomic.AddUint64(&ap.gasInPool, ^uint64(intrinsicGas-1))
 		//del actions in destination map
 		ap.deleteAccountDestinationActions(act)
 	}
 }
 
 // deleteAccountDestinationActions just for destination map
 func (ap *actPool) deleteAccountDestinationActions(acts ...action.SealedEnvelope) {
+	// add lock
+	ap.accountDesActs.mu.Lock()
+	defer ap.accountDesActs.mu.Unlock()
 	for _, act := range acts {
 		hash, err := act.Hash()
 		if err != nil {
 			log.L().Debug("Skipping action due to hash error", zap.Error(err))
 			continue
 		}
-		desAddress, ok := act.Destination()
-		if ok {
-			dst := ap.accountDesActs[desAddress]
+		if desAddress, ok := act.Destination(); ok {
+			dst := ap.accountDesActs.acts[desAddress]
 			if dst != nil {
 				delete(dst, hash)
 			}
+			if len(dst) == 0 {
+				delete(ap.accountDesActs.acts, desAddress)
+			}
 		}
 	}
 }
 
-// updateAccount updates queue's status and remove invalidated actions from pool if necessary
-func (ap *actPool) updateAccount(sender string) {
-	queue := ap.accountActs[sender]
-	acts := queue.UpdateQueue(queue.PendingNonce())
-	if len(acts) > 0 {
-		ap.removeInvalidActs(acts)
-	}
-	// Delete the queue entry if it becomes empty
-	if queue.Empty() {
-		delete(ap.accountActs, sender)
-	}
-}
+func (ap *actPool) enqueue(ctx context.Context, act action.SealedEnvelope) error {
+	var errChan = make(chan error)
 
-func (ap *actPool) reset() {
-	timer := ap.timerFactory.NewTimer("reset")
-	defer timer.End()
-
-	// Remove confirmed actions in actpool
-	ap.removeConfirmedActs()
-	for from, queue := range ap.accountActs {
-		// Reset pending balance for each account
-		addr, _ := address.FromString(from)
-		state, err := accountutil.AccountState(ap.sf, addr)
-		if err != nil {
-			log.L().Error("Error when resetting actpool state.", zap.Error(err))
-			return
-		}
-		queue.SetPendingBalance(state.Balance)
+	ap.jobQueue[ap.allocatedWorker(act.SrcPubkey().Address())] <- workerJob{ctx, act, errChan}
 
-		// Reset pending nonce and remove invalid actions for each account
-		confirmedNonce := state.Nonce
-		pendingNonce := confirmedNonce + 1
-		queue.SetPendingNonce(pendingNonce)
-		ap.updateAccount(from)
+	for {
+		select {
+		case <-ctx.Done():
+			close(errChan)
+			return ctx.Err()
+		case ret := <-errChan:
+			return ret
+		}
```

Comment:
this for loop will hold `enqueue`, so `Add()` won't be able to return?

Reply:
yes